bodo/io/iceberg/filter_conversion.py

```diff
@@ -0,0 +1,178 @@
+"""Convert Iceberg expressions to PyArrow expressions and scalars for filtering.
+This module imports pyicberg so it shouldn't be imported unless PyIceberg is installed."""
+
+from typing import Any
+
+import pyarrow as pa
+import pyarrow.compute as pc
+from pyiceberg.expressions import BoundTerm
+from pyiceberg.expressions.literals import Literal
+from pyiceberg.expressions.visitors import BoundBooleanExpressionVisitor
+
+
+# Adapted from https://github.com/apache/iceberg-python/blob/3070e7a2b8d681cd02f753b3a46e6ff1b27b76cf/pyiceberg/io/pyarrow.py#L702
+class _ConvertToArrowExpressionStringAndScalar(
+    BoundBooleanExpressionVisitor[tuple[str, list[tuple[str, Any]]]]
+):
+    """Visitor to convert a bound Iceberg expression to a PyArrow expression string and associated scalars.
+    Creates expr_filter_f_str and filter_scalars for use in other functions like get_iceberg_pq_dataset."""
+
+    def make_names_unique(
+        self,
+        left_child: tuple[str, list[tuple[str, Any]]],
+        right_child: tuple[str, list[tuple[str, Any]]],
+    ) -> tuple[tuple[str, list[tuple[str, Any]]], tuple[str, list[tuple[str, Any]]]]:
+        """Make sure the names in the left and right child results are unique, we assume that all names are of form fn where n is a number and
+        they are increasing with index."""
+        left_names = {name for name, _ in left_child[1]}
+        right_names = {name for name, _ in right_child[1]}
+
+        # If there is no overlap, we can just return the children as is
+        if left_names.isdisjoint(right_names):
+            return left_child, right_child
+
+        # This is a simple implementation that just sets the name to the lowest unused number
+        max_left_name = left_child[1][-1][0] if left_child[1] else "f0"
+        next_name = int(max_left_name[1:]) + 1
+        right_rename_map = {
+            name: f"f{next_name + i}" for i, (name, _) in enumerate(right_child[1])
+        }
+
+        # Rename the right child names in the scalar tuple
+        right_child_scalars = [
+            (right_rename_map[name], value) for name, value in right_child[1]
+        ]
+        # Rename the right child names in the expression string
+        right_child_expr = right_child[0]
+        for old_name, new_name in right_rename_map.items():
+            right_child_expr = right_child_expr.replace(old_name, new_name)
+        right_child = (right_child_expr, right_child_scalars)
```

Comment:
This function looks potentially error prone to me but can't think of a better way.